src/index.spec.ts

```diff
@@ -1,12 +1,176 @@
+  test('Should return an array of numbers without duplicates', () => {
+    expect(removeDuplicatesFromArray([1, 1, 3, 4, 4, 6])).toEqual([1, 3, 4, 6]);
+  });
+  test('Should return return the same array', () => {
+    expect(removeDuplicatesFromArray([1])).toEqual([1]);
+  });
+  test('Should return an error message requesting user to enter a valid array', () => {
+    expect(() => {
+      removeDuplicatesFromArray(randomString);
+    }).toThrowError('please provide an array of numbers or strings');
+  });
+  test('Should return an error message', () => {
+    expect(() => {
+      removeDuplicatesFromArray(emptyArray);
+    }).toThrow;
+  });
+});
+
+describe('Takes an email string and throws an error if the email is not authorized, else returns random product', () => {
+  test('Should return a product object with ID', () => {
+    expect(createRandomProduct('clark@kent.com')).toHaveProperty('id');
+    expect(createRandomProduct('clark@kent.com')).toHaveProperty('name');
+    expect(createRandomProduct('clark@kent.com')).toHaveProperty('description');
+    expect(createRandomProduct('clark@kent.com')).toHaveProperty('price');
+    expect(createRandomProduct('clark@kent.com')).toHaveProperty('tags');
+  });
+  test('Should return an error', () => {
+    expect(() => {
+      createRandomProduct('bruce@wayne.com');
+    }).toThrowError('You are not allowed to create products');
+  });
+  test('Should return an error', () => {
+    expect(() => {
+      createRandomProduct('diana@themyscira.com');
+    }).toThrowError('You are not allowed to create products');
+  });
+});
+
+describe('Takes a product object and returns an error if object properties are not valid, else returns the product object with its ID', () => {
+  test('Should return an error', () => {
+    expect(() => {
+      createProduct(invalidProduct);
+    }).toThrow();
+  });
+  test('Should return an error', () => {
+    expect(() => {
+      createProduct(emptyObject);
+    }).toThrow();
+  });
+  test('Should return a product object with ID', () => {
+    expect(createProduct(validProduct)).toHaveProperty('id');
+    expect(createProduct(validProduct)).toHaveProperty('name');
+    expect(createProduct(validProduct)).toHaveProperty('description');
+    expect(createProduct(validProduct)).toHaveProperty('price');
+    expect(createProduct(validProduct)).toHaveProperty('tags');
+  });
+});
+
+const server = setupServer(
+  rest.get('https://swapi.dev/api/planets', (req, res, ctx) => {
+    return res(
+      ctx.status(200),
+      ctx.json({
+        count: 60,
+        next: 'https://swapi.dev/api/planets/?page=2',
+      }),
+    );
+    // return res(
+    //   ctx.status(400),
+    //   ctx.json({
+    //     errorMessage: 'unable to make request',
+    //   }),
+    // );
+  }),
+);
+
+beforeAll(() => server.listen());
+afterAll(() => server.close());
+afterEach(() => server.resetHandlers());
```

Comment:
Generally speaking, jest globals such as `before<something>()` and `after<something()>` are better located at the start or end of the respective tested scope.

In this case, these globals should be located inside the starwars describe. 

`beforeAll(() => server.listen());` before the first test block. And `afterAll(() => server.close())` and `afterEach(() => server.resetHandlers())` after the last test block.

These globals outside any describe block would be OK if the influence all the tests in the file.

Reply:
😮 yes, that makes sense. I also saw a different pattern where these were in its own setup file and later imported. Thanks for the suggestion!